src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs

```diff
@@ -0,0 +1,174 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Net.Http;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3;
+using Microsoft.AspNetCore.Server.Kestrel.Core.WebTransport;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.WebTransport;
+
+/// <summary>
+/// Controls the WebTransport session and keeps track of all the streams.
+/// </summary>
+internal class WebTransportSession : IWebTransportSession
+{
+    private readonly Dictionary<long, WebTransportStream> _openStreams = new();
+    private readonly ConcurrentQueue<WebTransportStream> _pendingStreams = new();
+    private readonly SemaphoreSlim _pendingAcceptStreamRequests = new(0);
+    private readonly Http3Connection _connection;
+    private readonly Http3Stream _controlStream = default!;
+    private bool _isClosing;
+
+    internal static string WebTransportProtocolValue => "webtransport";
+    internal static string SecPrefix => "sec-webtransport-http3-";
+    internal static string VersionHeaderPrefix => $"{SecPrefix}draft";
+
+    // Order is important as we choose the first supported
+    // version (preferenc dereases as index increases)
+    internal static readonly IEnumerable<string> SuppportedWebTransportVersions = new string[]
+    {
+        $"{VersionHeaderPrefix}02"
+    };
+
+    long IWebTransportSession.SessionId => _controlStream.StreamId;
+
+    internal WebTransportSession(Http3Connection connection, Http3Stream controlStream)
+    {
+        _connection = connection;
+        _controlStream = controlStream;
+        _isClosing = false;
+    }
+
+    void IWebTransportSession.Abort()
+    {
+        Abort(new(), Http3ErrorCode.NoError);
+    }
+
+    internal void OnClientConnectionClosed()
+    {
+        if (_isClosing)
+        {
+            return;
+        }
+
+        _isClosing = true;
+
+        lock (_openStreams)
+        {
+            foreach (var stream in _openStreams)
+            {
+                stream.Value.Close();
+            }
+
+            _openStreams.Clear();
+        }
+
+        _pendingStreams.Clear();
+    }
+
+    internal void Abort(ConnectionAbortedException exception, Http3ErrorCode error)
+    {
+        if (_isClosing)
+        {
+            return;
+        }
+
+        _isClosing = true;
+        lock (_openStreams)
+        {
+            _controlStream.Abort(exception, error);
+            foreach (var stream in _openStreams)
+            {
+                stream.Value.Abort(exception);
```

Comment:
Careful passing an exception to multiple streams. If it gets thrown more than once it can corrupt the stack trace. You might need to clone it. Look at the Http2 cleanup logic that bulk closes streams.